torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3827,28 +3827,47 @@ struct to_ir {
   Value* emitTupleSlice(
       const SourceRange& loc,
       const NamedValue& tuple_val,
-      const NamedValue& beg_val,
-      const at::optional<NamedValue>& end_val) {
+      const std::vector<at::optional<NamedValue>>& tuple_args) {
     auto tuple_type = tuple_val.value(*graph)->type()->expect<TupleType>();
-    int64_t beg = getAdjTupleIndex(
-        loc,
-        tuple_type,
-        getSliceInd(beg_val.value(*graph), loc),
-        /*allow_out_of_bounds*/ true);
-    int64_t end;
     int64_t tuple_len = tuple_type->elements().size();
+    auto beg_val = tuple_args[0];
+    auto end_val = tuple_args[1];
+    auto step = tuple_args[2];
+    // first figure out the step sign to know if we want to flip the start or
+    // end. For example, if we have x[::-3], the start should be len(x) - 1 and
+    // end should be 0
+    int64_t step_size = 1;
+    if (step) {
+      auto val = toIValue(step->value(*graph));
+      TORCH_CHECK(val->isInt(), "Step size should always be a valid integer");
+      step_size = val->to<int64_t>();
+    }
+
+    int64_t beg = 0;
+    if (beg_val) {
+      beg = getAdjTupleIndex(
+          loc, tuple_type, getSliceInd(beg_val->value(*graph), loc), true);
+    } else {
+      // we want to flip the beg if we are slicing in the opposite direction
+      beg = step_size > 0 ? 0 : tuple_len;
```

Comment:
Should it be `0` or `tuple_len` - 1? IIRC, `begin` is inclusive so it must be a valid index.

Also none of your test cases cover this case (negative step, begin not specified)